holoviews/ipython/magics.py

```diff
@@ -337,25 +344,48 @@ def output(self, line, cell=None):
             print("\nFor help with the %output magic, call %output?")
             return
 
-        restore_copy = OrderedDict(OutputMagic.options.items())
+        # Make backup of previous options
+        prev_backend = Store.current_backend
+        prev_renderer = Store.renderers[prev_backend]
+        prev_mode = prev_renderer.mode
+        prev_params = {k: v for k, v in prev_renderer.get_param_values()
+                       if k in self.render_params}
+        prev_restore = dict(OutputMagic.options)
         try:
-            options = OrderedDict(OutputMagic.options.items())
-            new_options = self.get_options(line, options, cell is None)
-            self._set_render_options(new_options)
+            # Process magic
+            new_options = self.get_options(line, {}, cell is None)
+
+            # Make backup of options on selected renderer
+            if 'backend' in new_options:
+                backend = new_options['backend']
+                if ':' not in backend:
+                    backend += ':default'
+            else:
+                backend = prev_backend+':'+prev_mode
+            renderer = Store.renderers[backend.split(':')[0]]
+            render_params = {k: v for k, v in renderer.get_param_values()
+                             if k in self.render_params}
+
+            # Set options on selected renderer and set display hook options
             OutputMagic.options = new_options
+            self._set_render_options(new_options, backend)
         except Exception as e:
-            self.update_options(options, {'backend': restore_copy['backend']})
-            OutputMagic.options = restore_copy
-            self._set_render_options(restore_copy)
+            # If setting options failed ensure they are reset
+            OutputMagic.options = prev_restore
+            self.set_backend(prev_backend)
             print('Error: %s' % str(e))
             print("For help with the %output magic, call %output?\n")
             return
 
         if cell is not None:
             self.shell.run_cell(cell, store_history=STORE_HISTORY)
-            self.update_options(options, {'backend': restore_copy['backend']})
-            OutputMagic.options = restore_copy
-            self._set_render_options(restore_copy)
+            # After cell magic restore previous options and restore
+            # temporarily selected renderer
+            OutputMagic.options = prev_restore
+            self._set_render_options(render_params, backend)
+            if backend.split(':')[0] != prev_backend:
+                self.set_backend(prev_backend)
+                self._set_render_options(prev_params, prev_backend+':'+prev_mode)
```

Comment:
I'm tempted to say something about context managers. Maybe for another day and another PR...